// hypersdk-client/dist/apiTransformers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.processAPIBlock = exports.processAPITransactionStatus = exports.processAPITxResult = void 0;
const utils_1 = require("@noble/hashes/utils");
const utils_2 = require("@noble/hashes/utils");
function processAPITxResult(response, marshaler) {
    return {
        ...response,
        result: response.result.map(output => marshaler.parseTyped((0, utils_1.hexToBytes)(output), "output")[0])
    };
}
exports.processAPITxResult = processAPITxResult;
function processAPITransactionStatus(response, marshaler) {
    const processedOutputs = response.outputs.map(output => marshaler.parseTyped((0, utils_1.hexToBytes)(output), "output")[0]);
    return {
        ...response,
        outputs: processedOutputs,
    };
}
exports.processAPITransactionStatus = processAPITransactionStatus;
function processAPIBlock(response, marshaler) {
    const processedTxs = response.block.block.txs.map(tx => ({
        ...tx,
        auth: {
            signer: (0, utils_2.bytesToHex)(new Uint8Array(tx.auth.signer)),
            signature: (0, utils_2.bytesToHex)(new Uint8Array(tx.auth.signature))
        }
    }));
    const processedResults = response.block.results.map(result => processAPITransactionStatus(result, marshaler));
    return {
        ...response.block,
        block: {
            ...response.block.block,
            txs: processedTxs,
        },
        results: processedResults,
    };
}
exports.processAPIBlock = processAPIBlock;
